src/core/models/dt_linearmodel.cc

```diff
@@ -0,0 +1,1096 @@
+  T last_epoch = nepochs - static_cast<T>(niterations) + 1;
+
+  size_t iteration_nrows = dt_X_train->nrows();
+  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * static_cast<T>(iteration_nrows));
+  size_t total_nrows = (niterations - 1) * iteration_nrows + last_iteration_nrows;
+  size_t iteration_end;
+
+  // If a validation set is provided, we adjust the `iteration_nrows`
+  // to correspond to the `nepochs_val` epochs. After each iteration, we calculate
+  // the loss on the validation dataset, and trigger early stopping
+  // if relative loss does not decrese by at least `val_error`.
+  bool validation = _notnan(nepochs_val);
+  T loss = T_NAN;    // This value is returned when validation is not enabled
+  T loss_old = T(0); // Value of `loss` for a previous iteraction
+  std::vector<T> loss_history;
+  const Column& target_col0_val = validation? dt_y_val->get_column(0)
+                                            : target_col0_train;  // whatever
+  colvec cols_val;
+
+  if (validation) {
+    cols_val = make_casted_columns<T>(dt_X_val, stype);
+    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val * static_cast<T>(iteration_nrows)));
+    niterations = total_nrows / iteration_nrows + (total_nrows % iteration_nrows > 0);
+    loss_history.resize(val_niters, 0.0);
+  }
+
+
+  // Mutex to update global feature importance information
+  std::mutex m;
+
+  // Calculate work amounts for full fit iterations, last fit iteration and
+  // validation.
+  size_t work_total = (niterations - 1) * get_work_amount(iteration_nrows);
+  work_total += get_work_amount(total_nrows - (niterations - 1) * iteration_nrows);
+  if (validation) work_total += niterations * get_work_amount(dt_X_val->nrows());
+
+  // Set work amount to be reported by the zero thread.
+  dt::progress::work job(work_total);
+  job.set_message("Fitting...");
+  NThreads nthreads = nthreads_from_niters(iteration_nrows, MIN_ROWS_PER_THREAD);
+
+  dt::parallel_region(nthreads,
+    [&]() {
+      // Each thread gets a private storage observations and feature importances.
+      tptr<T> x = tptr<T>(new T[nfeatures]);
+      tptr<T> fi = tptr<T>(new T[nfeatures]());
+
+      for (size_t iter = 0; iter < niterations; ++iter) {
+        size_t iteration_start = iter * iteration_nrows;
+        iteration_end = (iter == niterations - 1)? total_nrows :
+                                                   (iter + 1) * iteration_nrows;
+        size_t iteration_size = iteration_end - iteration_start;
+
+        // Training.
+        dt::nested_for_static(iteration_size, ChunkSize(MIN_ROWS_PER_THREAD), [&](size_t i) {
+          size_t ii = (iteration_start + i) % dt_X_train->nrows();
+          U target;
+          bool isvalid = target_col0_train.get_element(ii, &target);
+
+          if (isvalid && _isfinite(target) && read_row(ii, cols, x)) {
+            // Loop over all the labels
+            for (size_t k = 0; k < label_ids_train.size(); ++k) {
```

Comment:
I don't think that's how multinomial regression is supposed to work... 
I'm more used to log-likelihood formulation, but basically in a multinomial regression the particular value of `y_i` in the current row determines which term will be added to the total likelihood that will in the end be maximized. So it's 
`log-likelihood = Sum[ log Pr(x_i | y_i) ]`. 
I mean, there shouldn't be a for-loop over all labels -- just one label, `y_i` is in play here.

Reply:
For the label `y_i` I train its binomial classifier on a positive, and for other classifiers I train on negative.